tiatoolbox/tools/stainextract.py

```diff
@@ -14,35 +14,207 @@
+        >>> extractor = RuifrokExtractor()
+        >>> stain_matrix = extractor.get_stain_matrix(img)
 
     """
 
     @staticmethod
     def get_stain_matrix(_):
+        """Get the pre-defined stain matrix.
+
+        Returns:
+            ndarray: pre-defined  stain matrix.
+
+        """
         return np.array([[0.65, 0.70, 0.29], [0.07, 0.99, 0.11]])
+
+
+class MacenkoExtractor:
+    """Get the stain matrix as defined in:
+
+    Macenko, Marc, et al. "A method for normalizing histology
+    slides for quantitative analysis." 2009 IEEE International
+    Symposium on Biomedical Imaging: From Nano to Macro. IEEE, 2009.
+
+    This class contains code inspired by StainTools
+    [https://github.com/Peter554/StainTools] written by Peter Byfield.
+
+    Examples:
+        >>> from tiatoolbox.tools.staiextract import MacenkoExtractor()
+        >>> extractor = MacenkoExtractor()
+        >>> stain_matrix = extractor.get_stain_matrix(img)
+
+    """
+
+    @staticmethod
+    def get_stain_matrix(img, luminosity_threshold=0.8, angular_percentile=99):
+        """Stain matrix estimation.
+
+        Args:
+            img (ndarray): input image used for stain matrix estimation
+            luminosity_threshold (float): threshold used for tissue area selection
+            angular_percentile (int):
+
+        Returns:
+            ndarray: estimated stain matrix.
+
+        """
+        img = img.astype("uint8")  # ensure input image is uint8
+        # convert to OD and ignore background
+        tissue_mask = get_luminosity_tissue_mask(
+            img, threshold=luminosity_threshold
+        ).reshape((-1,))
+        img_od = convert_RGB2OD(img).reshape((-1, 3))
+        img_od = img_od[tissue_mask]
+
+        # eigenvectors of cov in OD space (orthogonal as cov symmetric)
+        _, V = np.linalg.eigh(np.cov(img_od, rowvar=False))
+
+        # the two principle eigenvectors
+        V = V[:, [2, 1]]
+
+        # make sure vectors are pointing the right way
+        if V[0, 0] < 0:
+            V[:, 0] *= -1
+        if V[0, 1] < 0:
+            V[:, 1] *= -1
+
+        # project on this basis.
+        That = np.dot(img_od, V)
+
+        # angular coordinates with repect to the prinicple, orthogonal eigenvectors
+        phi = np.arctan2(That[:, 1], That[:, 0])
+
+        # min and max angles
+        min_phi = np.percentile(phi, 100 - angular_percentile)
+        max_phi = np.percentile(phi, angular_percentile)
+
+        # the two principle colors
+        v1 = np.dot(V, np.array([np.cos(min_phi), np.sin(min_phi)]))
+        v2 = np.dot(V, np.array([np.cos(max_phi), np.sin(max_phi)]))
+
+        # order of H&E - H first row
+        if v1[0] > v2[0]:
+            he = np.array([v1, v2])
+        else:
+            he = np.array([v2, v1])
+
+        normalised_rows = he / np.linalg.norm(he, axis=1)[:, None]
+        return normalised_rows
+
+
+class VahadaneExtractor:
+    """Get the stain matrix as defined in:
```

Comment:
Same as for Ruifork and Macenko, first line should end with a period.

Reply:
Fixed